File: 153-find-minimum-in-rotated-sorted-array.go
func findMin(nums []int) int {
    return helper(nums, 0, len(nums)-1)
}

func helper(nums []int, st, ed int) int {
    if (st == ed) {
        return nums[st]
    }
    mid := (st + ed) / 2
    if nums[mid] > nums[ed]{
        return helper(nums, mid+1, ed)
    }
    return helper(nums, st, mid)
}

File: bwu/binary_search/153-find-minimum-in-rotated-sorted-array.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        l, r = 0, n - 1
        while l + 1 < r:
            mid = (l + r) / 2
            if nums[r] < nums[l]:
                if nums[mid] > nums[l]:
                    l = mid
                else:
                    r = mid
            else:
                return nums[l]
                
        return min(nums[l], nums[r])
